GitHub Issue #4886

jbottigliero opened the pull request:
- Adds "Environment" (--env) to stats output.

**What kind of change does this PR introduce?**

feature

**Did you add tests for your changes?**

Added an `--env` option stats test.

**If relevant, link to documentation update:**

N/A

**Summary**

Implements #4886 

**Does this PR introduce a breaking change?**

N/A

**Other information**

sokra commented on test/binCases/stats/env/webpack.config.js:
```diff
@@ -0,0 +1,11 @@
+
+module.exports = {
+	entry: path.resolve(__dirname, "./index"),
+	stats: {
```
Could you clone this test but pass `stats.env: false`? I guess this doesn't work.

jbottigliero replied:
[Test Added](https://github.com/webpack/webpack/pull/4903/files#diff-586f1614cbbb91da6f7a9186efaba135).

sokra commented on bin/webpack.js:
```diff
@@ -188,6 +188,10 @@ function processOptions(options) {
 		outputOptions.context = firstOptions.context;
 
+	ifArg("env", function(environment) {
+		outputOptions.env = environment;
```
``` js
if(outputOptions.env)
```

Could be added here to make it work.

jbottigliero replied:
Maybe I'm misunderstanding – but I've updated the check in [`lib/Stats.js`](https://github.com/webpack/webpack/pull/4903/files#diff-baf371f5446dc9bc4b41022587e05b48R206) to handle the bool case(s) that were a result of the preset object. This value (`--env`) seems to be a bit of a special case since the majority of the other properties/values are being read from the `compilation` or defined on the `stats` instance directly. I considered both options but this implementation seemed the least intrusive... I can certainly make a change either way.

sokra commented on test/statsCases/preset-normal/expected.txt:
```diff
@@ -1,4 +1,5 @@
 Hash: c5a6856b43905ae12f17
+Environment (--env): None
```
It would be better to omit this line if the `--env` is not used. There is no benefit in adding it.

sokra commented on test/binCases/stats/no-env/test.js:
```diff
@@ -0,0 +1,10 @@
+"use strict";
```
make sure to add a `test.opts` file for this testcase too.

jbottigliero replied:
This one was meant to act as the negative case.

sokra commented on test/binCases/stats/env/test.opts:
```diff
@@ -0,0 +1 @@
+--env foobar
```
Try something more complex like: `--env.debug --env.number 123`

sokra commented on lib/Stats.js:
```diff
@@ -498,6 +501,11 @@ class Stats {
 		}
+		if(obj.env) {
+			colors.normal("Environment (--env): ");
+			colors.bold(obj.env);
```
I guess a JSON.stringify would be great here if the env is an object.

jbottigliero replied:
Updated to use `JSON.stringify`... I did throw in an argument for for spacing. It can certainly be removed, but seems like it might be helpful when passing more complex objects:

`--env.production --env.platform 'web' --env.version 2`
![screen shot 2017-05-22 at 9 43 32 pm](https://cloud.githubusercontent.com/assets/694253/26336629/3c1da6d0-3f39-11e7-8211-f493b45e88f4.png)

Maybe with this use case in mind, it would make sense to print this after `Public Path: ` with a double newline?

 Worth mentioning more "basic" configurations would just be the single line:

`--env production`
![screen shot 2017-05-22 at 10 01 01 pm](https://cloud.githubusercontent.com/assets/694253/26336816/3971409e-3f3a-11e7-8224-da1f8bdb6a06.png)

jbottigliero commented:
@sokra – I've added, and updated a few test to better cover the  various `--env` argument `stats.env` permutations. 

In doing so it also seemed reasonable to alter how the actual `--env` argument was being passed to `lib/Stats`. Rather than adding the value to the compilations (like `hash`), I'm now assigning it to the `Stats` instance directly, similar to the `timings` implementation. This change allows the `stats.env` option fallback to be `true` which seems to better align with the desired use case.

If this approach is approved I'll be sure to open a corresponding PR to the documentation for `stats.env`.

welll commented:
@jbottigliero let me know if I can help you ;)

jbottigliero commented:
I've updated the base branch to `webpack-2` as this is what the original change-set was developed against. However, the changes in _this_ PR are a bit problematic. The change in `bin/webpack.js` was setting the `outputOptions.env` using the `--env` arguments directly. That means the `lib/Stats.js` component would output the `Environment` section even when you explicitly set `stats: { env: false }` but passed a truth-y `env` argument.  I don't think this is expected, or intended behavior.

I've cleaned up the implementation in a new branch. Here is _that_ diff: https://github.com/webpack/webpack/compare/webpack-2...jbottigliero:issue-4886-2.4.1

Now, although the issue is tagged as `webpack-2`, I've taken a look at the latest master (currently `3.6.0`) and it looks like a similar solution would be viable (looks like the `argv.env` handling was just abstracted a bit, so it's not a clean-merge). ...but, I'm not exactly sure how the webpack team would like me to propose those changes (assuming this is desired for 3+). Get this merged and cherry-pick it on to the latest as a new PR? Create an entirely new branch and set of commits?
